File: Main Drivetrain Code/MainDrive.java
package org.firstinspires.ftc.teamcode;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;
import com.qualcomm.robotcore.eventloop.opmode.TeleOp;
import com.qualcomm.robotcore.hardware.DcMotor;
import com.qualcomm.robotcore.hardware.Servo;
import org.firstinspires.ftc.robotcore.external.JavaUtil;

@TeleOp(name = "OUR_MAIN_DRIVE1 (Blocks to Java)")
public class OUR_MAIN_DRIVE1 extends LinearOpMode {

  private Servo armservo;
  private DcMotor frontleft;
  private DcMotor rearleft;
  private DcMotor leftslide;
  private DcMotor frontright;
  private DcMotor rearright;
  private Servo claw;
  private DcMotor winch;

  /**
   * This function is executed when this Op Mode is selected from the Driver Station.
   */
  @Override
  public void runOpMode() {
    int cor_toggle;
    double denominator;
    float y;
    float rx;
    double x;
    boolean b_pressed;
    int n;

    armservo = hardwareMap.get(Servo.class, "arm servo");
    frontleft = hardwareMap.get(DcMotor.class, "front left");
    rearleft = hardwareMap.get(DcMotor.class, "rear left");
    leftslide = hardwareMap.get(DcMotor.class, "left slide");
    frontright = hardwareMap.get(DcMotor.class, "front right");
    rearright = hardwareMap.get(DcMotor.class, "rear right");
    claw = hardwareMap.get(Servo.class, "claw");
    winch = hardwareMap.get(DcMotor.class, "winch");

    armservo.setDirection(Servo.Direction.FORWARD);
    frontleft.setDirection(DcMotor.Direction.REVERSE);
    rearleft.setDirection(DcMotor.Direction.REVERSE);
    leftslide.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
    leftslide.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
    leftslide.setZeroPowerBehavior(DcMotor.ZeroPowerBehavior.BRAKE);
    frontleft.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
    frontright.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
    rearleft.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
    rearright.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
    frontleft.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
    frontright.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
    rearleft.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
    rearright.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
    cor_toggle = 0;
    waitForStart();
    armservo.setPosition(0.5);
    while (opModeIsActive()) {
      // Remember, Y stick value is reversed
      y = -gamepad1.left_stick_y;
      // Factor to counteract imperfect strafing
      x = gamepad1.left_stick_x * 1.1;
      rx = gamepad1.right_stick_x;
      // Denominator is the largest motor power (absolute value) or 1.
      // This ensures all powers maintain the same ratio, but only if one is outside of the range [-1, 1].
      denominator = JavaUtil.maxOfList(JavaUtil.createListWith(JavaUtil.sumOfList(JavaUtil.createListWith(Math.abs(y), Math.abs(x), Math.abs(rx))), 1));
      // Make sure your ID's match your configuration
      frontleft.setPower((y + x + rx) / denominator);
      rearleft.setPower(((y - x) + rx) / denominator);
      frontright.setPower(((y - x) - rx) / denominator);
      rearright.setPower(((y + x) - rx) / denominator);
      // all left side is reversed becasue of mototr mounting
      // Left slide code
      if (gamepad2.dpad_up || gamepad2.dpad_down) {
        leftslide.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
      }
      if (gamepad2.dpad_down) {
        leftslide.setPower(-0.6);
      } else {
        if (gamepad2.dpad_up) {
          leftslide.setPower(1);
        } else {
          leftslide.setPower(0);
        }
      }
      if (gamepad2.y && armservo.getPosition() <= 0.5) {
        armservo.setPosition(armservo.getPosition() + 0.02);
      }
      if (gamepad2.a && armservo.getPosition() >= 0.1 && armservo.getPosition() <= 1) {
        armservo.setPosition(armservo.getPosition() - 0.02);
      }
      if (gamepad2.left_bumper) {
        claw.setPosition(1);
      }
      if (gamepad2.right_bumper) {
        claw.setPosition(0.6);
      }
      if (gamepad2.b) {
        if (!b_pressed) {
          if (cor_toggle == 0) {
            cor_toggle = 1;
          } else {
            cor_toggle = 0;
          }
          b_pressed = true;
        }
      } else {
        b_pressed = false;
      }
      if (gamepad2.dpad_up || gamepad2.dpad_down) {
        n = leftslide.getCurrentPosition();
      }
      if (cor_toggle == 0) {
        if (!(gamepad2.dpad_up || gamepad2.dpad_down)) {
          if (Math.abs(n - leftslide.getCurrentPosition()) > 25) {
            leftslide.setTargetPosition(n);
            leftslide.setMode(DcMotor.RunMode.RUN_TO_POSITION);
            leftslide.setTargetPosition(n);
            leftslide.setPower(1);
          }
        }
      }
      if (gamepad1.a) {
        armservo.setPwmDisable();
        winch.setPower(1);
      }
      if (gamepad1.b) {
        armservo.setPwmDisable();
        winch.setPower(-1);
      }
      telemetry.addData("arm position", armservo.getPosition());
      telemetry.addData("cor toggle", cor_toggle);
      telemetry.addData("n", n);
      telemetry.addData("leftslide actual pos", leftslide.getCurrentPosition());
      telemetry.addData("front left pos", frontleft.getCurrentPosition());
      telemetry.addData("front right pos", frontright.getCurrentPosition());
      telemetry.addData("rear left pos", rearleft.getCurrentPosition());
      telemetry.addData("rear right pos", rearright.getCurrentPosition());
      telemetry.update();
    }
  }
}

File: Main Drivetrain Code/MainDrive.py
from com.qualcomm.robotcore.eventloop.opmode import LinearOpMode, TeleOp
from com.qualcomm.robotcore.hardware import DcMotor, Servo
from org.firstinspires.ftc.robotcore.external import JavaUtil

@TeleOp(name="OUR_MAIN_DRIVE1 (Blocks to Python)")
class OUR_MAIN_DRIVE1(LinearOpMode):

    def runOpMode(self):
        cor_toggle = 0
        n = 0
        b_pressed = False

        # Initialize hardware
        armservo = self.hardwareMap.servo.get("arm servo")
        frontleft = self.hardwareMap.dcMotor.get("front left")
        rearleft = self.hardwareMap.dcMotor.get("rear left")
        leftslide = self.hardwareMap.dcMotor.get("left slide")
        frontright = self.hardwareMap.dcMotor.get("front right")
        rearright = self.hardwareMap.dcMotor.get("rear right")
        claw = self.hardwareMap.servo.get("claw")
        winch = self.hardwareMap.dcMotor.get("winch")

        # Servo and motor setup
        armservo.direction = Servo.Direction.FORWARD
        frontleft.direction = DcMotor.Direction.REVERSE
        rearleft.direction = DcMotor.Direction.REVERSE

        leftslide.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        leftslide.mode = DcMotor.RunMode.RUN_WITHOUT_ENCODER
        leftslide.zeroPowerBehavior = DcMotor.ZeroPowerBehavior.BRAKE

        motors = [frontleft, frontright, rearleft, rearright]
        for motor in motors:
            motor.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
            motor.mode = DcMotor.RunMode.RUN_WITHOUT_ENCODER

        self.waitForStart()

        armservo.position = 0.5

        while self.opModeIsActive():
            # Joystick controls
            y = -self.gamepad1.left_stick_y  # Y stick value is reversed
            x = self.gamepad1.left_stick_x * 1.1  # Strafing adjustment
            rx = self.gamepad1.right_stick_x

            # Normalize motor powers
            denominator = max(abs(y) + abs(x) + abs(rx), 1)
            frontleft_power = (y + x + rx) / denominator
            rearleft_power = (y - x + rx) / denominator
            frontright_power = (y - x - rx) / denominator
            rearright_power = (y + x - rx) / denominator

            # Set motor powers
            frontleft.power = frontleft_power
            rearleft.power = rearleft_power
            frontright.power = frontright_power
            rearright.power = rearright_power

            # Left slide control
            if self.gamepad2.dpad_up or self.gamepad2.dpad_down:
                leftslide.mode = DcMotor.RunMode.RUN_WITHOUT_ENCODER

            if self.gamepad2.dpad_down:
                leftslide.power = -0.6
            elif self.gamepad2.dpad_up:
                leftslide.power = 1
            else:
                leftslide.power = 0

            # Arm servo control
            if self.gamepad2.y and armservo.position <= 0.5:
                armservo.position += 0.02
            if self.gamepad2.a and 0.1 <= armservo.position <= 1:
                armservo.position -= 0.02

            # Claw servo control
            if self.gamepad2.left_bumper:
                claw.position = 1
            if self.gamepad2.right_bumper:
                claw.position = 0.6

            # Toggle correction mode
            if self.gamepad2.b:
                if not b_pressed:
                    cor_toggle = 1 - cor_toggle
                    b_pressed = True
            else:
                b_pressed = False

            # Left slide position correction
            if self.gamepad2.dpad_up or self.gamepad2.dpad_down:
                n = leftslide.currentPosition

            if cor_toggle == 0 and not (self.gamepad2.dpad_up or self.gamepad2.dpad_down):
                if abs(n - leftslide.currentPosition) > 25:
                    leftslide.targetPosition = n
                    leftslide.mode = DcMotor.RunMode.RUN_TO_POSITION
                    leftslide.power = 1

            # Winch and arm servo disable
            if self.gamepad1.a:
                armservo.pwmDisable()
                winch.power = 1
            if self.gamepad1.b:
                armservo.pwmDisable()
                winch.power = -1

            # Telemetry data
            self.telemetry.addData("arm position", armservo.position)
            self.telemetry.addData("cor toggle", cor_toggle)
            self.telemetry.addData("n", n)
            self.telemetry.addData("leftslide actual pos", leftslide.currentPosition)
            self.telemetry.addData("front left pos", frontleft.currentPosition)
            self.telemetry.addData("front right pos", frontright.currentPosition)
            self.telemetry.addData("rear left pos", rearleft.currentPosition)
            self.telemetry.addData("rear right pos", rearright.currentPosition)
            self.telemetry.update()
